G-Research.OpenTelemetry.Processor.Partial/PartialActivityProcessor.cs

```diff
@@ -125,72 +153,113 @@ private ILoggerFactory CreateLoggerFactory()
             builder.AddOpenTelemetry(options =>
             {
                 options.IncludeScopes = true;
-                options.AddProcessor(logProcessor);
+                options.AddProcessor(_logProcessor);
                 options.SetResourceBuilder(ResourceBuilder.CreateEmpty()
                     .AddAttributes(ParentProvider.GetResource().Attributes));
             });
         });
     }
 
-    public override void OnEnd(Activity data)
+    private Dictionary<string, object> GetHeartbeatLogRecordAttributes() => new()
+    {
+        ["partial.event"] = "heartbeat",
+        ["partial.frequency"] = $"{_heartbeatIntervalMilliseconds}ms",
+        ["partial.body.type"] = "json/v1",
+    };
+
+    private static Dictionary<string, object> GetStopLogRecordAttributes() => new()
     {
-        using (logger.Value.BeginScope(GetStopLogRecordAttributes()))
+        ["partial.event"] = "stop",
+        ["partial.body.type"] = "json/v1",
+    };
+
+    private void ProcessQueues()
+    {
+        var triggers = new WaitHandle[] { _shutdownTrigger };
+
+        while (true)
         {
-            logger.Value.LogInformation(
-                SpecHelper.Json(new TracesData(data, TracesData.Signal.Stop)));
+            try
+            {
+                WaitHandle.WaitAny(triggers, _processIntervalMilliseconds);
+
+                ProcessDelayedHeartbeatActivities();
+                ProcessReadyHeartbeatActivities();
+            }
+            catch (ObjectDisposedException)
+            {
+                return;
+            }
         }
+    }
+
+    private void ProcessDelayedHeartbeatActivities()
+    {
+        while (_delayedHeartbeatActivities.TryPeek(out var span) &&
+               span.InitialHeartbeatTime <= DateTime.UtcNow)
+        {
+            _delayedHeartbeatActivities.TryDequeue(out span);
 
-        endedActivities.Enqueue(new KeyValuePair<ActivitySpanId, Activity>(data.SpanId, data));
+            if (_activeActivities.TryGetValue(span.SpanId, out _))
+            {
+                _readyHeartbeatActivities.Enqueue((span.SpanId,
+                    DateTime.UtcNow.AddMilliseconds(_heartbeatIntervalMilliseconds)));
+            }
+        }
```

Comment:
I _think_ we should send a log event before enqueuing the span to the _readyHeartbeatActivities queue?